File: src/api/authApi.ts
import { invoke } from "@tauri-apps/api/core";
import { UserInformationDto } from "../types/backend/dto/userInformationDto";
import SignUpRequest from "../types/backend/dto/signUpRequest";

export function signIn(
	username: string,
	password: string,
): Promise<UserInformationDto> {
	return invoke("sign_in", {
		username,
		password,
	});
}

export function signUp(request: SignUpRequest): Promise<UserInformationDto> {
	return invoke("sign_up", {
		request,
	});
}

export function signOut(): Promise<boolean> {
	return invoke("sign_out");
}

export function isSignedIn(): Promise<boolean> {
	return invoke("is_signed_in");
}

export function verifyUserEmail(verificationCode: string): Promise<void> {
	return invoke("verify_user_email", { verificationCode });
}

export function resendEmailVerificationCode(): Promise<void> {
	return invoke("resend_email_verification_code");
}

export function updatePassword(
	oldPassword: string,
	newPassword: string,
): Promise<void> {
	return invoke("update_password", { oldPassword, newPassword });
}

File: src-tauri/src/backend/api/auth_api.rs
use std::sync::Arc;

use crate::{
    backend::{
        backend_dto::{UpdatePasswordDto, UserInformationDto},
        clients::brainy_backend_client::BrainyBackendClient,
        dto::sign_up_request_dto::SignUpRequestDto,
        services::authenticator::Authenticator,
    },
    common::api_error::ApiError,
};
use injector::injector::Injector;
use tauri::State;

#[tauri::command]
pub async fn sign_in(
    injector: State<'_, Arc<Injector>>,
    username: String,
    password: String,
) -> Result<UserInformationDto, ApiError> {
    let scope = injector.start_scope();
    let dto = scope
        .resolve::<dyn Authenticator>()
        .await
        .sign_in(username, password)
        .await?;
    Ok(dto)
}

#[tauri::command]
pub async fn sign_up(
    injector: State<'_, Arc<Injector>>,
    request: SignUpRequestDto,
) -> Result<UserInformationDto, ApiError> {
    let scope = injector.start_scope();
    let dto = scope
        .resolve::<dyn Authenticator>()
        .await
        .sign_up(request)
        .await?;
    Ok(dto)
}

#[tauri::command]
pub async fn sign_out(injector: State<'_, Arc<Injector>>) -> Result<(), ApiError> {
    let scope = injector.start_scope();
    scope
        .resolve::<dyn Authenticator>()
        .await
        .sign_out()
        .await?;
    Ok(())
}

#[tauri::command]
pub async fn is_signed_in(injector: State<'_, Arc<Injector>>) -> Result<bool, ApiError> {
    let scope = injector.start_scope();
    Ok(scope
        .resolve::<dyn BrainyBackendClient>()
        .await
        .is_signed_in()?)
}

#[tauri::command]
pub async fn verify_user_email(
    injector: State<'_, Arc<Injector>>,
    verification_code: String,
) -> Result<(), ApiError> {
    let scope = injector.start_scope();
    scope
        .resolve::<dyn BrainyBackendClient>()
        .await
        .verify_user_email(verification_code)
        .await?;
    Ok(())
}

#[tauri::command]
pub async fn resend_email_verification_code(
    injector: State<'_, Arc<Injector>>,
) -> Result<(), ApiError> {
    let scope = injector.start_scope();
    scope
        .resolve::<dyn BrainyBackendClient>()
        .await
        .resend_email_verification_code()
        .await?;
    Ok(())
}

#[tauri::command]
pub async fn update_password(
    injector: State<'_, Arc<Injector>>,
    old_password: String,
    new_password: String,
) -> Result<(), ApiError> {
    let scope = injector.start_scope();
    scope
        .resolve::<dyn BrainyBackendClient>()
        .await
        .update_password(UpdatePasswordDto {
            old_password,
            new_password,
        })
        .await?;
    Ok(())
}
